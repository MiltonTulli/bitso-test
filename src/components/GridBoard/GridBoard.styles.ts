import { createUseStyles } from 'react-jss';
import theme, { Theme } from '../../theme';

export default createUseStyles<Theme>({
  root: {
    display: 'flex',
    justifyContent: 'center',
    padding: theme.spacing(2)
  }
});
