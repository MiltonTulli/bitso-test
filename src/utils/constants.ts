export const MATRIX_SIZES = ['5x5', '10x5', '10x10', '15x10', '15x15', '20x20'];
export const IS_MOBILE_BREAK = 600;
