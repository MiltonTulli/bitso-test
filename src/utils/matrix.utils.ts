import { ICell, IMatrix } from '../interfaces';

// export utils -----------------
/**
 * @function createMatrix
 * @description Create a matrix (array of arrays) given cols & rows
 * @param {Number} cols - Num of columns (x)
 * @param {Number} rows - Num of rows (y)
 */
export const createMatrix = (cols: number, rows: number): ICell[][] => {
  let matrix: ICell[][] = [];
  for (let y = 0; y < rows; y++) {
    matrix[y] = [];
    for (let x = 0; x < cols; x++) {
      matrix[y][x] = {
        id: `${x}-${y}`,
        active: false
      };
    }
  }
  return matrix;
};
/**
 * @function parseStringId
 * @description Convert string with 'N-N' format and return number for both coordenates
 * @param {String} id - 'N-N' format
 * @returns {Array} with [ x, y ] format where x & y are numbers
 */
export const parseStringId = (id: string): number[] => {
  if (!/^\d+-\d+$/.test(id)) return [];
  const coordenate = id.split('-');
  const x = Number(coordenate[0]);
  const y = Number(coordenate[1]);
  return [x, y];
};

/**
 * @function getIslands
 * @description - Get matrix groups ('islands')
 * @param {ICell[]} activeCells - Active cells
 * @returns {Array<Array>} - Return an array with arrays for each group ("island") found
 */
export const getIslands = (activeCells: ICell[]): string[][] => {
  const activeIds: string[] = activeCells.map(cell => cell.id);
  let result: string[][] = [];
  activeIds.forEach(p1 => {
    const groups = result.filter(group => group.some(p2 => isFamily(p1, p2)));
    if (groups.length) {
      result = result.filter(r => !groups.some(g => g === r));
      return result.push([...groups.flat(), p1]);
    }
    result.push([p1]);
  });
  return result;
};

export const calculateCellSize = (
  rowCount: number,
  windowWidth: number,
  spacingX: number
): number => {
  const space = windowWidth - spacingX;
  return Number(space / rowCount);
};

/**
 * @function isFamily
 * @param {String} idA - id to compre
 * @param {String} idB - id to compare
 * @returns {Boolean} - boolean value containing if id 'A' is family of id 'B'
 */
export const isFamily = (idA: string, idB: string): boolean => {
  if (idA === idB) return true;
  if (!idA) return false;
  const [x, y] = parseStringId(idA);
  switch (idB) {
    case `${x}-${y - 1}`:
    case `${x}-${y + 1}`:
    case `${x - 1}-${y}`:
    case `${x + 1}-${y}`:
      return true;
    default:
      return false;
  }
};

export const idHasFamilyOnArr = (id: string, arr: string[]): boolean => {
  if (!Array.isArray(arr)) return false;
  for (let idB of arr) {
    if (isFamily(id, idB)) {
      return true;
    }
  }
  return false;
};

export const getActiveCells = (matrix: IMatrix): ICell[] => {
  let active: ICell[] = [];
  for (let row of matrix) {
    for (let cell of row) {
      if (cell.active) {
        active.push(cell);
      }
    }
  }
  return active;
};
/**
 * Returns boolean whether two points are diagonally aligned;
 * @param { String } idA
 * @param { String } idB
 */
export const isDiagonal = (idA: string, idB: string): boolean => {
  const [x, y] = parseStringId(idA);
  return [
    `${x - 1}-${+y - 1}`,
    `${x - 1}-${+y + 1}`,
    `${x + 1}-${+y + 1}`,
    `${x + 1}-${+y - 1}`
  ].some(e => idB === e);
};
/**
 * Returns the edges of a diagonal
 * @param { String } idA
 * @param { String } idB
 */
export const getAristas = (idA: string, idB: string) => {
  const [Ax, Ay] = parseStringId(idA);
  const [Bx, By] = parseStringId(idB);
  const simbolX = Bx - Ax;
  const simbolY = By - Ay;
  return [`${Ax + simbolX}-${Ay}`, `${Ax}-${Ay + simbolY}`];
};
