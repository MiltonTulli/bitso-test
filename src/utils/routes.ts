export default {
  ROOT: '/',
  BOARD: '/board'
};
