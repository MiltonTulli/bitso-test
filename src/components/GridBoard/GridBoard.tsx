import React from 'react';
import clsx from 'classnames';
import { Cell } from '..';
import styles from './GridBoard.styles';
import { ICell, IMatrix } from '../../interfaces';
import { useDimensions } from '..';
import { calculateCellSize } from '../../utils/matrix.utils';

interface GridBoardProps {
  matrix: IMatrix;
  handleToggle: (id: string) => void;
  columns: number;
}

export const GridBoard: React.FunctionComponent<GridBoardProps> = (props: GridBoardProps) => {
  const { matrix, handleToggle, columns } = props;
  const { width } = useDimensions();
  const size = calculateCellSize(columns, width, 20 * 4);
  const classes = styles();
  return (
    <div className={clsx(classes.root)}>
      <table className="" cellSpacing="0">
        <tbody>
          {/* // rows */}
          {matrix &&
            matrix.length > 0 &&
            matrix.map((rowArr, y) => {
              return (
                <tr data-testid="row" key={rowArr[0].id.split('-')[1]}>
                  {/* // cols */}
                  {matrix[y].map((colCell: ICell) => {
                    return (
                      <td key={colCell.id}>
                        <Cell
                          data-testid="col"
                          size={size}
                          position={colCell.id}
                          active={colCell.active}
                          handleToggle={handleToggle}
                        />
                      </td>
                    );
                  })}
                </tr>
              );
            })}
        </tbody>
      </table>
    </div>
  );
};
