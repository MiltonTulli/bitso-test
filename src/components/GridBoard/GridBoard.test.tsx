import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import { GridBoard } from './GridBoard';
import { createMatrix } from '../../utils/matrix.utils';

describe('<GridBoard>', () => {
  let props;
  let x = 10;
  let y = 5;

  beforeEach(() => {
    props = {
      matrix: createMatrix(x, y),
      handleToggle: jest.fn(),
      columns: 5
    };
  });

  it('Should match snapshot', () => {
    const wrapper = renderer.create(<GridBoard {...props} />).toJSON();
    expect(wrapper).toMatchSnapshot();
  });

  it('Should render correct num of rows', () => {
    const wrapper = shallow(<GridBoard {...props} />);
    const rowEl = wrapper.find('[data-testid="row"]');
    expect(rowEl.length).toEqual(y);
  });

  it('Should render correct num of cols', () => {
    const wrapper = shallow(<GridBoard {...props} />);
    const colEl = wrapper.find('[data-testid="col"]');
    expect(colEl.length).toEqual(y * x);
  });
});
