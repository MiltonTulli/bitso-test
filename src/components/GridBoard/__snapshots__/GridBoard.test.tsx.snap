// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GridBoard> Should match snapshot 1`] = `
<div
  className="root-0-2-1"
>
  <table
    cellSpacing="0"
    className=""
  >
    <tbody>
      <tr
        data-testid="row"
      >
        <td>
          <div
            className="root-0-2-2 root-d0-0-2-5 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d1-0-2-6 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d2-0-2-7 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d3-0-2-8 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d4-0-2-9 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d5-0-2-10 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d6-0-2-11 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d7-0-2-12 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d8-0-2-13 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d9-0-2-14 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
      </tr>
      <tr
        data-testid="row"
      >
        <td>
          <div
            className="root-0-2-2 root-d10-0-2-15 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d11-0-2-16 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d12-0-2-17 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d13-0-2-18 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d14-0-2-19 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d15-0-2-20 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d16-0-2-21 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d17-0-2-22 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d18-0-2-23 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d19-0-2-24 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
      </tr>
      <tr
        data-testid="row"
      >
        <td>
          <div
            className="root-0-2-2 root-d20-0-2-25 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d21-0-2-26 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d22-0-2-27 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d23-0-2-28 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d24-0-2-29 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d25-0-2-30 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d26-0-2-31 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d27-0-2-32 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d28-0-2-33 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d29-0-2-34 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
      </tr>
      <tr
        data-testid="row"
      >
        <td>
          <div
            className="root-0-2-2 root-d30-0-2-35 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d31-0-2-36 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d32-0-2-37 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d33-0-2-38 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d34-0-2-39 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d35-0-2-40 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d36-0-2-41 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d37-0-2-42 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d38-0-2-43 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d39-0-2-44 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
      </tr>
      <tr
        data-testid="row"
      >
        <td>
          <div
            className="root-0-2-2 root-d40-0-2-45 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d41-0-2-46 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d42-0-2-47 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d43-0-2-48 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d44-0-2-49 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d45-0-2-50 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d46-0-2-51 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d47-0-2-52 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d48-0-2-53 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
        <td>
          <div
            className="root-0-2-2 root-d49-0-2-54 inactive-0-2-4"
            data-testid="cell"
            onClick={[Function]}
          />
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
